sigstore/_internal/merkle.py

```diff
@@ -1,9 +1,115 @@
 """
 Utilities for verifying proof-of-inclusion within Rekor's Merkle Tree.
+
+This code is based off Google's Trillian Merkle Tree implementation which Cosign uses to validate
+Rekor entries.
+
+The data format for the Merkle tree nodes is described in IETF's RFC 6962.
 """
 
-from sigstore._internal.rekor import RekorInclusionProof
+import base64
+import hashlib
+import struct
+from typing import List, Tuple
 
+from sigstore._internal.rekor import RekorEntry, RekorInclusionProof
 
-def verify_merkle_inclusion(inclusion_proof: RekorInclusionProof) -> None:
+
+class InvalidInclusionProofError(Exception):
     pass
+
+
+LEAF_HASH_PREFIX = 0
+NODE_HASH_PREFIX = 1
+
+
+def _decomp_inclusion_proof(index: int, size: int) -> Tuple[int, int]:
```

Comment:
Some of this is pretty cryptic. I've basically translated directly from the Trillian Go source code. Not sure if I should try to simplify this or rewrite it.